File: android/app/src/main/java/blog/nextlab/echo/data/MessageEnvelope.kt
package blog.nextlab.echo.data

import blog.nextlab.echo.model.CallOutcome
import blog.nextlab.echo.model.MediaId
import blog.nextlab.echo.model.MessageContent
import blog.nextlab.echo.model.MessageText
import blog.nextlab.echo.model.StickerId
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi
import org.json.JSONException
import org.json.JSONObject

/**
 * 暗号化された封の中に、実際に入るもの。
 *
 * 最初の版は*本文*（1行の文字列）だけを封じ、それ以外はサーバーから読める Firestore の
 * ドキュメントに残していた。文字だけを封じている間はそれでよかったが、写真を送った
 * 瞬間に成り立たなくなる。id もサイズも形も**サムネイルも**封の外に出る。32pxの
 * サムネイルは小さいが、写真そのものには違いない。写真の説明文だけ暗号化して写真は
 * 平文で送るのは、暗号化ではなく装飾。
 *
 * なので封じる単位は、種類を問わず**content 全体**にして、小さな JSON 1つとして書く。
 * 外のドキュメントに残るのは経路に要るものだけ（誰が、いつ、封がされていること）。
 *
 * メディアの鍵がここを通るのは、保管庫のファイルが専用の鍵で暗号化されていて、その鍵を
 * 相手の端末へ届ける必要があるから。メッセージのドキュメントに入れると、データベースを
 * 読める者が全部の写真を取って開けることになり、保管庫を作った意味が消える。封の中なら、
 * そのメッセージを読める端末にしか届かない。保管庫には、運用者を含む誰にとっても
 * 無意味なバイト列だけが残る。
 *
 * これができる前に書かれたものは、封の中身が素の文字列だった。[open] は解釈できない
 * 平文を文字として扱うので、そういうメッセージも開き続ける。代償は失敗する JSON 解析
 * 1回で、それが「古いメッセージも読める」と「古いメッセージは消えた」の差になる。
 * 履歴を捨ててよい研究用のビルドでも、読めるメッセージを黙って失うのは、この形式が
 * どこまで信用できるかについて間違ったことを教える。
 */
@OptIn(ExperimentalEncodingApi::class)
object MessageEnvelope {

    /** 封じる平文として [content] を書き出す。 */
    fun seal(content: MessageContent): String = when (content) {
        is MessageContent.Text -> JSONObject()
            .put(TYPE, TYPE_TEXT)
            .put(BODY, content.body.value)
            .toString()

        is MessageContent.Image -> JSONObject()
            .put(TYPE, TYPE_IMAGE)
            .put(MEDIA_ID, content.mediaId.value)
            .put(WIDTH, content.width)
            .put(HEIGHT, content.height)
            .put(BYTES, content.byteCount)
            .put(THUMBNAIL, Base64.encode(content.thumbnail))
            .apply { content.mediaKey?.let { put(MEDIA_KEY, Base64.encode(it)) } }
            .apply { content.originalId?.let { put(ORIGINAL_ID, it.value) } }
            .apply { content.originalKey?.let { put(ORIGINAL_KEY, Base64.encode(it)) } }
            .apply { content.originalBytes?.let { put(ORIGINAL_BYTES, it) } }
            .apply { content.originalMime?.let { put(ORIGINAL_MIME, it) } }
            .toString()

        is MessageContent.Video -> JSONObject()
            .put(TYPE, TYPE_VIDEO)
            .put(MEDIA_ID, content.mediaId.value)
            .put(WIDTH, content.width)
            .put(HEIGHT, content.height)
            .put(DURATION_MS, content.durationMs)
            .put(BYTES, content.byteCount)
            .put(SEALED_BYTES, content.sealedBytes)
            .put(THUMBNAIL, Base64.encode(content.thumbnail))
            .apply { content.mediaKey?.let { put(MEDIA_KEY, Base64.encode(it)) } }
            .toString()

        is MessageContent.Sticker -> JSONObject()
            .put(TYPE, TYPE_STICKER)
            .put(STICKER_ID, content.stickerId.value)
            .toString()

        // 通話の記録は「通話があった」と言うだけで、鍵の要る要素は無い。それでも封じる。
        // 「どのメッセージが暗号化する価値が無かったか」自体、外から見えてよいものではない。
        is MessageContent.Call -> JSONObject()
            .put(TYPE, TYPE_CALL)
            .put(CALL_VIDEO, content.video)
            .put(CALL_OUTCOME, content.outcome.name.lowercase())
            .put(CALL_SECONDS, content.seconds)
            .toString()

        is MessageContent.Locked -> error("cannot seal an already-sealed message")
        MessageContent.Retracted -> error("cannot seal a retracted message")
    }

    /**
     * 復号された平文を content に戻す。
     *
     * 例外は投げない。解釈できないメッセージは新しい版が書いたものか壊れたもので、
     * それで落ちるスレッドは、1行描けないスレッドより悪い。null は
     * 「この端末にはこれが何か分からない」という意味。
     */
    fun open(plaintext: String): MessageContent? {
        val json = try {
            JSONObject(plaintext)
        } catch (_: JSONException) {
            // 封ができる前に書かれたもの。平文そのものが本文だった。
            return MessageContent.Text(MessageText(plaintext))
        }

        return when (json.optString(TYPE)) {
            TYPE_TEXT -> MessageContent.Text(MessageText(json.optString(BODY)))

            TYPE_IMAGE -> MessageContent.Image(
                mediaId = MediaId(json.optString(MEDIA_ID)),
                width = json.optInt(WIDTH),
                height = json.optInt(HEIGHT),
                thumbnail = json.optString(THUMBNAIL).let {
                    if (it.isEmpty()) ByteArray(0) else decode(it)
                },
                byteCount = json.optInt(BYTES),
                mediaKey = json.optString(MEDIA_KEY).ifEmpty { null }?.let(::decode),
                originalId = json.optString(ORIGINAL_ID).ifEmpty { null }?.let(::MediaId),
                originalKey = json.optString(ORIGINAL_KEY).ifEmpty { null }?.let(::decode),
                originalBytes = json.optInt(ORIGINAL_BYTES).takeIf { it > 0 },
                originalMime = json.optString(ORIGINAL_MIME).ifEmpty { null },
            )

            TYPE_VIDEO -> MessageContent.Video(
                mediaId = MediaId(json.optString(MEDIA_ID)),
                width = json.optInt(WIDTH),
                height = json.optInt(HEIGHT),
                durationMs = json.optLong(DURATION_MS),
                thumbnail = json.optString(THUMBNAIL).let {
                    if (it.isEmpty()) ByteArray(0) else decode(it)
                },
                byteCount = json.optInt(BYTES),
                sealedBytes = json.optLong(SEALED_BYTES),
                mediaKey = json.optString(MEDIA_KEY).ifEmpty { null }?.let(::decode),
            )

            TYPE_STICKER -> MessageContent.Sticker(
                StickerId(json.optString(STICKER_ID)),
            )

            TYPE_CALL -> MessageContent.Call(
                video = json.optBoolean(CALL_VIDEO),
                outcome = CallOutcome.entries
                    .firstOrNull { it.name.equals(json.optString(CALL_OUTCOME), ignoreCase = true) }
                    ?: CallOutcome.Completed,
                seconds = json.optInt(CALL_SECONDS),
            )

            // 封ではない普通の JSON オブジェクト。誰かがそう打っただけ。解釈はできたので
            // 上の分岐には落ちないが、これもただの文字。
            else -> MessageContent.Text(MessageText(plaintext))
        }
    }

    private fun decode(value: String): ByteArray =
        try {
            Base64.decode(value)
        } catch (_: IllegalArgumentException) {
            // 途中で切れているか壊れている。空のサムネイルは仮画像になるだけだが、
            // 例外を投げるとスレッドの残り全部を失う。
            ByteArray(0)
        }

    private const val TYPE = "t"
    private const val TYPE_TEXT = "text"
    private const val TYPE_IMAGE = "image"
    private const val TYPE_VIDEO = "video"
    private const val TYPE_STICKER = "sticker"
    private const val TYPE_CALL = "call"

    private const val BODY = "b"
    private const val MEDIA_ID = "id"
    private const val MEDIA_KEY = "k"
    private const val WIDTH = "w"
    private const val HEIGHT = "h"
    private const val BYTES = "n"
    private const val THUMBNAIL = "th"
    private const val ORIGINAL_ID = "oid"
    private const val ORIGINAL_KEY = "ok"
    private const val ORIGINAL_BYTES = "on"
    private const val ORIGINAL_MIME = "om"
    private const val DURATION_MS = "ms"
    private const val SEALED_BYTES = "sn"
    private const val STICKER_ID = "s"
    private const val CALL_VIDEO = "v"
    private const val CALL_OUTCOME = "o"
    private const val CALL_SECONDS = "sec"
}

File: ios/RinowaCore/Sources/RinowaCore/MessageEnvelope.swift
import Foundation

/// 暗号化された封の中に、実際に入るもの。
///
/// `MessageEnvelope.kt` の Swift 側。両方が research/vectors/formats.json に縛られ、
/// 仕様は docs/WIRE_FORMATS.md。
///
/// 本文だけでなく content 全体を封じる理由: 文字しか無かった頃は本文だけで足りたが、
/// 写真を送った瞬間に足りなくなる。id もサイズも**サムネイルも**封の外に出るから。
/// 32px のサムネイルは小さいが、写真そのものには違いない。写真の説明文だけ暗号化して
/// 写真を平文で送るのは、暗号化ではない。
public enum MessageContent: Equatable {
    case text(String)
    case image(Image)
    case video(Video)
    case sticker(String)
    case call(Call)

    public struct Image: Equatable {
        public let mediaId: String
        public let width: Int
        public let height: Int
        public let byteCount: Int
        public let thumbnail: Data
        public let mediaKey: Data?
        public let originalId: String?
        public let originalKey: Data?
        public let originalBytes: Int?
        public let originalMime: String?

        public init(
            mediaId: String,
            width: Int,
            height: Int,
            byteCount: Int,
            thumbnail: Data,
            mediaKey: Data? = nil,
            originalId: String? = nil,
            originalKey: Data? = nil,
            originalBytes: Int? = nil,
            originalMime: String? = nil
        ) {
            self.mediaId = mediaId
            self.width = width
            self.height = height
            self.byteCount = byteCount
            self.thumbnail = thumbnail
            self.mediaKey = mediaKey
            self.originalId = originalId
            self.originalKey = originalKey
            self.originalBytes = originalBytes
            self.originalMime = originalMime
        }
    }

    public struct Video: Equatable {
        public let mediaId: String
        public let width: Int
        public let height: Int
        public let durationMs: Int64
        public let byteCount: Int
        /// **暗号化後**のオブジェクトの大きさ。範囲で読む再生側が最初に必要とする。
        public let sealedBytes: Int64
        public let thumbnail: Data
        public let mediaKey: Data?

        public init(
            mediaId: String,
            width: Int,
            height: Int,
            durationMs: Int64,
            byteCount: Int,
            sealedBytes: Int64,
            thumbnail: Data,
            mediaKey: Data? = nil
        ) {
            self.mediaId = mediaId
            self.width = width
            self.height = height
            self.durationMs = durationMs
            self.byteCount = byteCount
            self.sealedBytes = sealedBytes
            self.thumbnail = thumbnail
            self.mediaKey = mediaKey
        }
    }

    public struct Call: Equatable {
        public enum Outcome: String {
            case completed, missed, declined, failed
        }

        public let video: Bool
        public let outcome: Outcome
        public let seconds: Int

        public init(video: Bool, outcome: Outcome, seconds: Int) {
            self.video = video
            self.outcome = outcome
            self.seconds = seconds
        }
    }
}

public enum MessageEnvelope {

    /// 封じる平文として `content` を書き出す。
    public static func seal(_ content: MessageContent) -> String {
        var object: [String: Any]

        switch content {
        case let .text(body):
            object = [Key.type: Kind.text, Key.body: body]

        case let .image(image):
            object = [
                Key.type: Kind.image,
                Key.mediaId: image.mediaId,
                Key.width: image.width,
                Key.height: image.height,
                Key.bytes: image.byteCount,
                Key.thumbnail: image.thumbnail.base64EncodedString()
            ]
            image.mediaKey.map { object[Key.mediaKey] = $0.base64EncodedString() }
            image.originalId.map { object[Key.originalId] = $0 }
            image.originalKey.map { object[Key.originalKey] = $0.base64EncodedString() }
            image.originalBytes.map { object[Key.originalBytes] = $0 }
            image.originalMime.map { object[Key.originalMime] = $0 }

        case let .video(video):
            object = [
                Key.type: Kind.video,
                Key.mediaId: video.mediaId,
                Key.width: video.width,
                Key.height: video.height,
                Key.durationMs: video.durationMs,
                Key.bytes: video.byteCount,
                Key.sealedBytes: video.sealedBytes,
                Key.thumbnail: video.thumbnail.base64EncodedString()
            ]
            video.mediaKey.map { object[Key.mediaKey] = $0.base64EncodedString() }

        case let .sticker(id):
            object = [Key.type: Kind.sticker, Key.stickerId: id]

        case let .call(call):
            object = [
                Key.type: Kind.call,
                Key.callVideo: call.video,
                Key.callOutcome: call.outcome.rawValue,
                Key.callSeconds: call.seconds
            ]
        }

        guard
            let data = try? JSONSerialization.data(withJSONObject: object),
            let text = String(data: data, encoding: .utf8)
        else {
            // 上の値はすべて String、Int、Bool、または Data 由来の String なので、
            // ここには来ない。落とさずに本文を返すことで、直列化の予想外がメッセージを
            // 道連れにしないようにする。
            if case let .text(body) = content { return body }
            return ""
        }
        return text
    }

    /// 復号された平文を content に戻す。例外は投げない。
    ///
    /// 解釈できないものは文字として返る。この形式ができる前に書かれたメッセージ
    /// （JSON ではない素の本文）が、いまも読めるのはそのため。
    public static func open(_ plaintext: String) -> MessageContent? {
        guard
            let data = plaintext.data(using: .utf8),
            let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return .text(plaintext)
        }

        switch object[Key.type] as? String {
        case Kind.text:
            return .text(object[Key.body] as? String ?? "")

        case Kind.image:
            return .image(
                MessageContent.Image(
                    mediaId: object[Key.mediaId] as? String ?? "",
                    width: intValue(object[Key.width]),
                    height: intValue(object[Key.height]),
                    byteCount: intValue(object[Key.bytes]),
                    thumbnail: decode(object[Key.thumbnail]) ?? Data(),
                    mediaKey: decode(object[Key.mediaKey]),
                    originalId: object[Key.originalId] as? String,
                    originalKey: decode(object[Key.originalKey]),
                    originalBytes: (object[Key.originalBytes] as? NSNumber)?.intValue,
                    originalMime: object[Key.originalMime] as? String
                )
            )

        case Kind.video:
            return .video(
                MessageContent.Video(
                    mediaId: object[Key.mediaId] as? String ?? "",
                    width: intValue(object[Key.width]),
                    height: intValue(object[Key.height]),
                    durationMs: Int64(intValue(object[Key.durationMs])),
                    byteCount: intValue(object[Key.bytes]),
                    sealedBytes: Int64(intValue(object[Key.sealedBytes])),
                    thumbnail: decode(object[Key.thumbnail]) ?? Data(),
                    mediaKey: decode(object[Key.mediaKey])
                )
            )

        case Kind.sticker:
            return .sticker(object[Key.stickerId] as? String ?? "")

        case Kind.call:
            let outcome = (object[Key.callOutcome] as? String)
                .flatMap(MessageContent.Call.Outcome.init(rawValue:)) ?? .completed
            return .call(
                MessageContent.Call(
                    video: object[Key.callVideo] as? Bool ?? false,
                    outcome: outcome,
                    seconds: intValue(object[Key.callSeconds])
                )
            )

        default:
            // 解釈できる JSON だが封ではない。誰かがそう打っただけ。やはりただの文字。
            return .text(plaintext)
        }
    }

    private static func decode(_ value: Any?) -> Data? {
        guard let text = value as? String, !text.isEmpty else { return nil }
        // 途中で切れているか壊れていれば、投げずに空として復号する。空のサムネイルは
        // 仮画像になるだけだが、例外はスレッドの残り全部を失う。
        return Data(base64Encoded: text)
    }

    private static func intValue(_ value: Any?) -> Int {
        (value as? NSNumber)?.intValue ?? 0
    }

    private enum Kind {
        static let text = "text"
        static let image = "image"
        static let video = "video"
        static let sticker = "sticker"
        static let call = "call"
    }

    private enum Key {
        static let type = "t"
        static let body = "b"
        static let mediaId = "id"
        static let mediaKey = "k"
        static let width = "w"
        static let height = "h"
        static let bytes = "n"
        static let thumbnail = "th"
        static let durationMs = "ms"
        static let sealedBytes = "sn"
        static let originalId = "oid"
        static let originalKey = "ok"
        static let originalBytes = "on"
        static let originalMime = "om"
        static let stickerId = "s"
        static let callVideo = "v"
        static let callOutcome = "o"
        static let callSeconds = "sec"
    }
}
